### ISSUE TITLE
log4js.shutdown(() => process.exit(1)) seems not work for appenders type 'log4js-logstash-tcp'

### ISSUE THREAD

**Issue by mathTaoTao** [log4js-node/log4js-node#1384]
Hi,
I am using appenders type 'log4js-logstash-tcp' to save log to Elasticsearch, and I used log4js.shutdown(() => process.exit(1)); , but it seems that the log was not saved to Elasticsearch when process exit. Any ideas?Please help me out.

**Comment by lamweili** [comment]
Hi, `log4js-logstash-tcp` is not maintained by us.

Nevertheless, the issue is at `log4js-logstash-tcp`, [index.js @ line 98-100](https://github.com/Aigent/log4js-logstash-tcp/blob/a9b11fbd2c01063cfede1e87f258bd9bde9d3ed1/index.js#L98-L100):

```js
    log.shutdown = function (cb) {
        cb();
    };
```

You might want to file an issue over there. Some form of graceful termination/flushing is needed within the `log.shutdown` function.

---

A sample of `@log4js-node/logstashudp`, [index.js @ line 57](https://github.com/log4js-node/logstashUDP/blob/c4b71e8657a366d9489367e78a7eac2cc5f1378b/src/index.js#L57):

```js
  log.shutdown = cb => udp.close(cb);
```

A sample of the shutdown function in [fileAppender @ line 120](https://github.com/log4js-node/log4js-node/blob/26dcec62f9677dceba57de8cd717ff91447781c7/lib/appenders/file.js#L120):

https://github.com/log4js-node/log4js-node/blob/26dcec62f9677dceba57de8cd717ff91447781c7/lib/appenders/file.js#L114-L121

**Comment by lamweili** [comment]
Hi, `log4js-logstash-tcp` is not maintained by us.

Nevertheless, the issue is at `log4js-logstash-tcp`, [index.js @ line 98-100](https://github.com/Aigent/log4js-logstash-tcp/blob/a9b11fbd2c01063cfede1e87f258bd9bde9d3ed1/index.js#L98-L100):

```js
    log.shutdown = function (cb) {
        cb();
    };
```

You might want to file an issue over there: https://github.com/Aigent/log4js-logstash-tcp
Some form of graceful termination/flushing is needed within the `log.shutdown` function.

---

A sample of `@log4js-node/logstashudp`, [index.js @ line 57](https://github.com/log4js-node/logstashUDP/blob/c4b71e8657a366d9489367e78a7eac2cc5f1378b/src/index.js#L57):

```js
  log.shutdown = cb => udp.close(cb);
```

A sample of the shutdown function in [fileAppender @ line 120](https://github.com/log4js-node/log4js-node/blob/26dcec62f9677dceba57de8cd717ff91447781c7/lib/appenders/file.js#L120):

https://github.com/log4js-node/log4js-node/blob/26dcec62f9677dceba57de8cd717ff91447781c7/lib/appenders/file.js#L114-L121

**Comment by lamweili** [comment]
I have created a PR https://github.com/Aigent/log4js-logstash-tcp/pull/7 over there.
But it's really up to the maintainers over there to get it released.

Perhaps you might want to file an issue there for tracking purposes.

**Comment by lamweili** [comment]
Migrated PR https://github.com/Aigent/log4js-logstash-tcp/pull/7 to PR https://github.com/Degola/log4js-logstash-tcp/pull/1.

**Comment by lamweili** [comment]
The maintainers for [`log4js-logstash-tcp`](https://github.com/Degola/log4js-logstash-tcp) has released 2.0.5 which fixed this issue.

I will close this. Let me know if you still face any issues.

Also, I'll mark as invalid because it is not related to this `log4js` repository.

**Comment by lamweili** [comment]
The maintainers for [`log4js-logstash-tcp`](https://github.com/Degola/log4js-logstash-tcp), @Degola, has released 2.0.5 which fixed this issue.

I will close this. Let me know if you still face any issues.

Also, I'll mark as invalid because it is not related to this `log4js` repository.
